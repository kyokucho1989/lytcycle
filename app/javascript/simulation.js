import * as d3 from "d3";

// データの初期値をロード

let routes;
let operators;
let facilities;
let link, node;
const routesInitial = [
  { source: 0, target: 1, l: 20, id: "root10" },
  { source: 1, target: 2, l: 20, id: "root11" },
  { source: 2, target: 0, l: 20, id: "root12" },
  { source: 0, target: 2, l: 20, id: "re_root10" },
  { source: 1, target: 0, l: 20, id: "re_root11" },
  { source: 2, target: 1, l: 20, id: "re_root12" },
];
let operatorsInitial = [{ name: "Alice" }];
const facilitiesInitial = [
  {
    index: 0,
    x: 230,
    y: 310,
    r: 10,
    id: "n0",
    processingTime: 1,
    type: "start",
    name: "startPoint",
  },
  {
    index: 1,
    x: 330,
    y: 60,
    r: 15,
    id: "n1",
    processingTime: 15,
    type: "machine",
    name: "machine-no-1",
    isProcessing: false,
    hasMaterial: false,
    storageSize: 0,
    processingEndTime: 0,
  },
  {
    index: 2,
    x: 430,
    y: 310,
    r: 10,
    id: "n2",
    processingTime: 1,
    type: "goal",
    name: "goalPoint",
  },
];

let simulation;

export async function drawLink(linksData, nodesData) {
  console.log("draw link");
  d3.select("#svg02").selectAll("line").remove();
  d3.select("#svg02").selectAll("circle").remove();

  link = d3
    .select("#svg02")
    .selectAll("line")
    .data(linksData)
    .enter()
    .append("line")
    .attr("stroke-width", 1)
    .attr("stroke", "black")
    .attr("id", function (d) {
      return d.id;
    });

  node = d3
    .select("#svg02")
    .selectAll("circle")
    .data(nodesData)
    .enter()
    .append("circle")
    .attr("r", function (d) {
      return d.r;
    })
    .attr("stroke", "black")
    .attr("fill", "LightSalmon")
    .attr("id", function (d) {
      return d.id;
    });

  const simurateSvg = document.getElementById("svg02");
  if (simurateSvg) {
    // シミュレーション描画
    simulation = d3
      .forceSimulation()
      .force("link", d3.forceLink().strength(0).iterations(1))
      .force("charge", d3.forceManyBody().strength(0))
      .force(
        "x",
        d3
          .forceX()
          .strength(0.01)
          .x(400 / 2)
      )
      .force(
        "y",
        d3
          .forceY()
          .strength(0.01)
          .y(100 / 2)
      );

    simulation.nodes(nodesData).on("tick", ticked);
    node
      .call(
        d3
          .drag()
          .on("start", dragstarted)
          .on("drag", dragged)
          .on("end", dragended)
      )
      .on("click", nodeClicked);

    simulation
      .force("link")
      .links(linksData)
      .id(function (d) {
        return d.index;
      });
  }
}

let facilityDialog, confirmBtn;
document.addEventListener("turbo:load", async () => {
  facilityDialog = document.getElementById("facilityDialog");
  confirmBtn = document.getElementById("confirmBtn");

  // ［確認］ボタンが既定でフォームを送信しないようにし、`close()` メソッドでダイアログを閉じ、"close" イベントを発生させる
  confirmBtn.addEventListener("click", (event) => {
    event.preventDefault(); // この偽フォームを送信しない
    let id = document.getElementById("hidden-id");
    let name = document.getElementById("name");
    let processingTime = document.getElementById("processingTime");

    let selectedFacility = facilities.find(
      (facility) => facility.id == id.value
    );
    selectedFacility.name = name.value;
    selectedFacility.processingTime = processingTime.value;

    facilityDialog.close();
  });
});
function nodeClicked() {
  let facilityForEdit = facilities.find((facility) => facility.id == this.id);
  console.log(facilityForEdit);
  setFacilityDataToModal(facilityForEdit);
  facilityDialog.showModal();
}

function setFacilityDataToModal(facility) {
  let id = document.getElementById("hidden-id");
  let name = document.getElementById("name");
  let processingTime = document.getElementById("processingTime");

  id.value = facility.id;
  name.value = facility.name;
  processingTime.value = facility.processingTime;
}

function dragstarted(event) {
  if (!event.active) simulation.alphaTarget(0.3).restart();
  event.subject.fx = event.subject.x;
  event.subject.fy = event.subject.y;
}

// Update the subject (dragged node) position during drag.
function dragged(event) {
  event.subject.fx = event.x;
  event.subject.fy = event.y;
}

function dragended(event) {
  if (!event.active) simulation.alphaTarget(0);
  event.subject.fx = null;
  event.subject.fy = null;
}

function ticked() {
  link
    .attr("x1", (d) => d.source.x)
    .attr("y1", (d) => d.source.y)
    .attr("x2", (d) => d.target.x)
    .attr("y2", (d) => d.target.y);

  node.attr("cx", (d) => d.x).attr("cy", (d) => d.y);
}

document.addEventListener("turbo:load", async () => {
  let simulationParameters = document.getElementById("simulation-data");
  console.log("シミュレーション");
  if (simulationParameters) {
    let simulationId = simulationParameters.dataset.id;
    if (simulationId == "") {
      // 初期値設定
      routes = routesInitial;
      operators = operatorsInitial;
      facilities = facilitiesInitial;
      console.log("初期値設定終了");
    } else {
      try {
        const response = await fetch("edit.json");
        if (!response.ok) {
          throw new Error(`レスポンスステータス: ${response.status}`);
        }
        const json = await response.json();
        console.log(json);
        routes = JSON.parse(json.routes);
        facilities = JSON.parse(json.facilities);
        operators = JSON.parse(json.operators);

        // routesのsourceとtargetをインデックスに修正
        routes.forEach((route) => {
          if (typeof route.source === "object") {
            route.source = route.source.index;
          }
          if (typeof route.target === "object") {
            route.target = route.target.index;
          }
        });

        drawLink(routes, facilities);
        return json;
      } catch (error) {
        console.error(error.message);
      }

      console.log("ロード完了");
    }

    // シミュレーション画面の描画
  }

  // データ整形
});
document.addEventListener("turbo:load", () => {
  const saveSimulationButton = document.getElementById("savesimulation");
  if (saveSimulationButton) {
    saveSimulationButton.addEventListener("click", (event) => {
      event.preventDefault();
      const form = document.getElementById("simulationForm");
      const formData = new FormData(form);
      formData.append("js_routes", JSON.stringify(routes));
      formData.append("js_operators", JSON.stringify(operators));
      formData.append("js_facilities", JSON.stringify(facilities));

      fetch(form.action, {
        method: "POST",
        headers: {
          "X-CSRF-Token": document
            .querySelector("meta[name='csrf-token']")
            .getAttribute("content"),
          Accept: "application/json",
        },
        body: formData,
      })
        .then((response) => response.json())
        .then((data) => {
          console.log("Success:", data);
          // 成功時の処理（ページ遷移など）
          window.location.href = data.location;
        })
        .catch((error) => {
          console.error("Error:", error);
          // エラー時の処理
        });
    });
  }
});
