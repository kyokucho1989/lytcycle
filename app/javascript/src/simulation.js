import * as d3 from "d3";
import { routes, operators, facilities } from "src/set_simulation_params";
// データの初期値をロード

export let link, node, simulation;
import { addFacility } from "src/set_simulation_params";
let facilityDialog, confirmBtn, routeDialog, routeConfirmBtn;
export async function drawLink(linksData, nodesData) {
  d3.select("#svg02").selectAll("line").remove();
  d3.select("#svg02").selectAll("circle").remove();

  link = d3
    .select("#svg02")
    .selectAll("line")
    .data(linksData)
    .enter()
    .append("line")
    .attr("stroke-width", 8)
    .attr("stroke", "black")
    .attr("id", function (d) {
      return d.id;
    });

  node = d3
    .select("#svg02")
    .selectAll("circle")
    .data(nodesData)
    .enter()
    .append("circle")
    .attr("r", function (d) {
      return d.r;
    })
    .attr("stroke", "black")
    .attr("fill", "LightSalmon")
    .attr("id", function (d) {
      return d.id;
    });

  const simurateSvg = document.getElementById("svg02");
  if (simurateSvg) {
    // シミュレーション描画
    simulation = d3
      .forceSimulation()
      .force("link", d3.forceLink().strength(0))
      .force("charge", d3.forceManyBody().strength(0))
      .force("x", null)
      .force("y", null);

    simulation.nodes(nodesData).on("tick", ticked);
    node
      .call(
        d3
          .drag()
          .on("start", dragstarted)
          .on("drag", dragged)
          .on("end", dragended)
      )
      .on("click", nodeClicked);
    link.on("click", linkClicked);
    simulation
      .force("link")
      .links(linksData)
      .id(function (d) {
        return d.index;
      });
  }
}

export function changeInactiveObject() {
  d3.select("#svg02").selectAll("line").on("click", null);
  d3.select("#svg02").selectAll("circle").on("click", null);
}

export function changeActiveObject() {
  d3.select("#svg02").selectAll("line").on("click", linkClicked);
  d3.select("#svg02").selectAll("circle").on("click", nodeClicked);
}

export function setClickEventToObject(object) {
  if (object.mainState == "running") {
    console.log("run");
    changeInactiveObject();
  } else {
    switch (object.subState) {
      case "select":
        d3.select("#svg02").on("click", null);
        console.log("select");
        break;
      case "add-operator":
        d3.select("#svg02").on("click", null);
        console.log("add-operator");
        break;
      case "add-facility":
        switchAddFacilityMode();
        break;
      case "delete":
        d3.select("#svg02").on("click", null);
        console.log("delete");
        break;
      case "link":
        d3.select("#svg02").on("click", null);
        console.log("link");
        break;
    }
  }

  d3.select("#svg02").selectAll("line").on("click", linkClicked);
  d3.select("#svg02").selectAll("circle").on("click", nodeClicked);
}

export function switchAddFacilityMode() {
  const svg = d3.select("#svg02");
  svg.on("click", null);
  svg.on("click", function (e) {
    const [x, y] = d3.pointer(e, this);
    addFacility([x, y]);
  });
}

export function setObjectparams(e, params, objects) {
  e.preventDefault(); // この偽フォームを送信しない
  let id = params.id;
  let selectedObject = objects.find((object) => object.id == id);
  for (const [key, value] of Object.entries(params)) {
    if (key != "id") {
      selectedObject[key] = value;
    }
  }
}

document.addEventListener("turbo:load", async () => {
  facilityDialog = document.getElementById("facilityDialog");
  confirmBtn = document.getElementById("confirmBtn");
  routeDialog = document.getElementById("route-dialog");
  routeConfirmBtn = document.getElementById("route-confirm-btn");
  if (confirmBtn) {
    // ［確認］ボタンが既定でフォームを送信しないようにし、`close()` メソッドでダイアログを閉じ、"close" イベントを発生させる
    confirmBtn.addEventListener("click", (e) => {
      // e.preventDefault();
      let params = {};
      params.id = document.getElementById("hidden-id").value;
      params.name = document.getElementById("name").value;
      params.processingTime = document.getElementById("processingTime").value;

      setObjectparams(e, params, facilities);
      facilityDialog.close();
    });
  }

  if (routeConfirmBtn) {
    // ［確認］ボタンが既定でフォームを送信しないようにし、`close()` メソッドでダイアログを閉じ、"close" イベントを発生させる
    routeConfirmBtn.addEventListener("click", (e) => {
      let params = {};
      params.id = document.getElementById("route-hidden-id").value;
      params.routeLength = document.getElementById("route-length").value;

      setObjectparams(e, params, routes);
      let selectedRoute = routes.find((route) => route.id == params.id);
      let pairRoute = routes.find(
        (route) =>
          route.target == selectedRoute.source &&
          route.source == selectedRoute.target
      );
      params.id = pairRoute.id;
      setObjectparams(e, params, routes);
      routeDialog.close();
    });
  }
});

function nodeClicked() {
  let facilityForEdit = facilities.find((facility) => facility.id == this.id);
  setFacilityDataToModal(facilityForEdit);
  facilityDialog.showModal();
}

function linkClicked() {
  const canvasController = document.querySelector(
    "[data-controller='simulationmode']"
  ).controller;
  console.log(canvasController);

  let routeForEdit = routes.find((route) => route.id == this.id);
  console.log(routeForEdit);
  setRouteDataToModal(routeForEdit);
  routeDialog.showModal();
}

export function setFacilityDataToModal(facility) {
  let id = document.getElementById("hidden-id");
  let name = document.getElementById("name");
  let processingTime = document.getElementById("processingTime");

  id.value = facility.id;
  name.value = facility.name;
  processingTime.value = facility.processingTime;
}

export function setRouteDataToModal(route) {
  let id = document.getElementById("route-hidden-id");
  let length = document.getElementById("route-length");

  id.value = route.id;
  length.value = route.routeLength;
}

function dragstarted(event) {
  if (!event.active) simulation.alphaTarget(0.3).restart();
  event.subject.fx = event.subject.x;
  event.subject.fy = event.subject.y;
}

// Update the subject (dragged node) position during drag.
function dragged(event) {
  event.subject.fx = event.x;
  event.subject.fy = event.y;
}

function dragended(event) {
  if (!event.active) simulation.alphaTarget(0);
  event.subject.fx = null;
  event.subject.fy = null;
}

function ticked() {
  link
    .attr("x1", (d) => d.source.x)
    .attr("y1", (d) => d.source.y)
    .attr("x2", (d) => d.target.x)
    .attr("y2", (d) => d.target.y);

  node.attr("cx", (d) => d.x).attr("cy", (d) => d.y);
}

document.addEventListener("turbo:load", () => {
  const saveSimulationButton = document.getElementById("savesimulation");
  if (saveSimulationButton) {
    saveSimulationButton.addEventListener("click", (event) => {
      event.preventDefault();
      const form = document.getElementById("simulationForm");
      const formData = new FormData(form);
      formData.append("js_routes", JSON.stringify(routes));
      formData.append("js_operators", JSON.stringify(operators));
      formData.append("js_facilities", JSON.stringify(facilities));

      fetch(form.action, {
        method: "POST",
        headers: {
          "X-CSRF-Token": document
            .querySelector("meta[name='csrf-token']")
            .getAttribute("content"),
          Accept: "application/json",
        },
        body: formData,
      })
        .then((response) => response.json())
        .then((data) => {
          console.log("Success:", data);
          // 成功時の処理（ページ遷移など）
          window.location.href = data.location;
        })
        .catch((error) => {
          console.error("Error:", error);
          // エラー時の処理
        });
    });
  }
});
