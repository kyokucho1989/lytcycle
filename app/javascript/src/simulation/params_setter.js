import { drawLink } from "../render";

export let routes, operators, facilities;

const routesInitial = [
  { source: "start", target: "n1", routeLength: 20, id: "root10" },
  { source: "n1", target: "goal", routeLength: 20, id: "root1-1" },
  { source: "goal", target: "start", routeLength: 20, id: "root-10" },
];
const operatorsInitial = [{ name: "Alice" }];
const facilitiesInitial = [
  {
    x: 430,
    y: 310,
    r: 10,
    id: "goal",
    processingTime: 1,
    type: "goal",
    name: "goal",
  },
  {
    x: 230,
    y: 310,
    r: 10,
    id: "start",
    processingTime: 1,
    type: "start",
    name: "start",
  },
  {
    x: 330,
    y: 60,
    r: 15,
    id: "n1",
    processingTime: 15,
    type: "machine",
    name: "machine-1",
    isProcessing: false,
    hasMaterial: false,
    storageSize: 0,
    processingEndTime: 0,
  },
];

export function deleteRoute() {
  const selectedRoute = routes.find((route) => route.id === this.id);
  if (
    selectedRoute.source.id === "goal" &&
    selectedRoute.target.id === "start"
  ) {
    alert("スタートとゴールの間のルートは削除できません。");
  } else {
    if (window.confirm("削除しますか？")) {
      routes = routes.filter((route) => route.id !== selectedRoute.id);
      drawLink(routes, facilities);
    }
  }
}

export function deleteFacility() {
  const selectedFacility = facilities.find(
    (facility) => facility.id === this.id
  );
  if (selectedFacility.id === "goal" || selectedFacility.id === "start") {
    alert("スタートとゴールは削除できません。");
    return;
  }
  const unConnectedRoutes = routes.filter(
    (route) =>
      route.target.id !== selectedFacility.id &&
      route.source.id !== selectedFacility.id
  );

  const unConnectedRoutesIds = unConnectedRoutes.map((route) => route.id);

  if (window.confirm("削除しますか？")) {
    facilities = facilities.filter(
      (facility) => facility.id !== selectedFacility.id
    );
    routes = routes.filter((route) => unConnectedRoutesIds.includes(route.id));
    drawLink(routes, facilities);
  }
}

export function addRoute(targetId, sourceId) {
  const targetNode = facilities.find((element) => element.id === targetId);
  const sourceNode = facilities.find((element) => element.id === sourceId);
  const idName = `root${sourceId}${targetId}`;
  const route = {
    source: sourceNode,
    target: targetNode,
    routeLength: 20,
    id: idName,
    index: routes.length,
  };

  const duplicatedLogic = (route) =>
    (route.target.id === targetId && route.source.id === sourceId) ||
    (route.target.id === sourceId && route.source.id === targetId);

  if (routes.some(duplicatedLogic)) {
    alert("すでにリンクが作成されています");
  } else {
    routes = routes.concat(route);
    drawLink(routes, facilities);
  }
  const selectedNodes = document.querySelectorAll("circle[selected]");

  [...selectedNodes].forEach((element) => {
    element.removeAttribute("selected");
  });
}

export function addFacility([x, y]) {
  const lastIdObject = facilities.find((el) => el.lastId);
  let lastId;
  if (lastIdObject === undefined) {
    lastId = facilities.length;
    facilities.unshift({ lastId: lastId });
  } else {
    lastId = lastIdObject.lastId;
  }

  lastId = lastId + 1;
  facilities[0] = { lastId: lastId, x: 0, y: 0 };
  const facility = {
    index: lastId,
    x: x,
    y: y,
    r: 15,
    id: `n${lastId}`,
    processingTime: 15,
    type: "machine",
    name: `machine-${lastId}`,
    isProcessing: false,
    hasMaterial: false,
    storageSize: 0,
    processingEndTime: 0,
  };

  facilities.push(facility);
  drawLink(routes, facilities);
}

document.addEventListener("turbo:load", async () => {
  const simulationParameters = document.getElementById("simulation-data");
  if (simulationParameters) {
    const simulationId = simulationParameters.dataset.id;

    if (simulationId === "") {
      routes = routesInitial;
      operators = operatorsInitial;
      facilities = facilitiesInitial;
      drawLink(routes, facilities);
    } else {
      try {
        const response = await fetch("edit.json");
        if (!response.ok) {
          throw new Error(`レスポンスステータス: ${response.status}`);
        }
        const json = await response.json();
        routes = JSON.parse(json.routes);
        facilities = JSON.parse(json.facilities);
        operators = JSON.parse(json.operators);

        routes.forEach((route) => {
          if (typeof route.source === "object") {
            route.source = route.source.id;
          }
          if (typeof route.target === "object") {
            route.target = route.target.id;
          }
        });

        drawLink(routes, facilities);
        return json;
      } catch (error) {
        console.error(error.message);
      }
    }
  }
});
