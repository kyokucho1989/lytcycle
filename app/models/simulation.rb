class Simulation < ApplicationRecord
  belongs_to :user
  validates :title, length: { maximum: 20 }, presence: true
end
