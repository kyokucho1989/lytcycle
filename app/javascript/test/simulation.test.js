// query utilities:
import {
  getByLabelText,
  getByText,
  getByTestId,
  queryByTestId,
  // Tip: all queries are also exposed on an object
  // called "queries" which you could import here as well
  waitFor,
} from "@testing-library/dom";
// adds special assertions like toHaveTextContent
import "@testing-library/jest-dom";

import * as d3 from "d3";
import { drawLink } from "../simulation.js";

test("SVG Snapshot Test", () => {
  document.body.innerHTML = '<div id="chart"></div>';

  const svg = d3
    .select("#chart")
    .append("svg")
    .attr("width", 100)
    .attr("height", 100);

  svg
    .append("circle")
    .attr("cx", 50)
    .attr("cy", 50)
    .attr("r", 40)
    .attr("fill", "blue");

  expect(document.body.innerHTML).toMatchSnapshot();
});

test("draw link and circle", async () => {
  const routesInitial = [
    { source: 0, target: 1, l: 20, id: "root10" },
    { source: 1, target: 2, l: 20, id: "root11" },
    { source: 2, target: 0, l: 20, id: "root12" },
    { source: 0, target: 2, l: 20, id: "re_root10" },
    { source: 1, target: 0, l: 20, id: "re_root11" },
    { source: 2, target: 1, l: 20, id: "re_root12" },
  ];

  const facilitiesInitial = [
    {
      index: 0,
      x: 230,
      y: 310,
      r: 10,
      id: "n0",
      processingTime: 1,
      type: "start",
      name: "startPoint",
    },
    {
      index: 1,
      x: 330,
      y: 60,
      r: 15,
      id: "n1",
      processingTime: 15,
      type: "machine",
      name: "machine-no-1",
      isProcessing: false,
      hasMaterial: false,
      storageSize: 0,
      processingEndTime: 0,
    },
    {
      index: 2,
      x: 430,
      y: 310,
      r: 10,
      id: "n2",
      processingTime: 1,
      type: "goal",
      name: "goalPoint",
    },
  ];

  await drawLink(routesInitial, facilitiesInitial);

  const div = document.createElement("div");
  div.innerHTML = `
    <svg id="svg02" xmlns="http://www.w3.org/2000/svg" width="700" height="800">
    </svg>
  `;

  document.body.innerHTML = '<svg id="svg02"></div>';

  expect(div).toMatchSnapShot();
});
