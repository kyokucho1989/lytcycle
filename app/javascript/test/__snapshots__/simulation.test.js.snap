// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw link and circle 1`] = `
<div>
  
    
  <svg
    height="800"
    id="svg02"
    width="700"
    xmlns="http://www.w3.org/2000/svg"
  >
    
    
    <line
      id="root10"
      stroke="black"
      stroke-width="8"
      x1="230"
      x2="330"
      y1="310"
      y2="60"
    />
    <line
      id="root11"
      stroke="black"
      stroke-width="8"
      x1="330"
      x2="430"
      y1="60"
      y2="310"
    />
    <line
      id="root12"
      stroke="black"
      stroke-width="8"
      x1="430"
      x2="230"
      y1="310"
      y2="310"
    />
    <line
      id="re_root10"
      stroke="black"
      stroke-width="8"
      x1="230"
      x2="430"
      y1="310"
      y2="310"
    />
    <line
      id="re_root11"
      stroke="black"
      stroke-width="8"
      x1="330"
      x2="230"
      y1="60"
      y2="310"
    />
    <line
      id="re_root12"
      stroke="black"
      stroke-width="8"
      x1="430"
      x2="330"
      y1="310"
      y2="60"
    />
    <circle
      cx="230"
      cy="310"
      fill="LightSalmon"
      id="n0"
      r="10"
      stroke="black"
    />
    <circle
      cx="330"
      cy="60"
      fill="LightSalmon"
      id="n1"
      r="15"
      stroke="black"
    />
    <circle
      cx="430"
      cy="310"
      fill="LightSalmon"
      id="n2"
      r="10"
      stroke="black"
    />
  </svg>
  
  
</div>
`;
